include/dali/core/exec/tasking/task.h

```diff
@@ -0,0 +1,522 @@
+ *
+ * A Task is a function-like object that can carry some dependencies and preconditions
+ * (via Succeed) and subscribe for results of other tasks (via Consume).
+ *
+ * A task function can produce no result (void), a scalar result (any type) or multiple results
+ * (iterable object or a tuple).
+ *
+ * Task lifecycle:
+ * 1. New
+ *    - the task is created
+ *    - the data dependencies are added
+ * 2. Pending
+ *    - the task has been submitted to a Scheduler via AddTask or AddSilentTask
+ * 3. Ready
+ *    - the task has been submitted and all its preconditions are met
+ * 4. Running
+ *    - the task has been Popped from the scheduler and is being run
+ * 5. Complete
+ *    - the task's payload has finished running
+ *
+ * A Task also manages its output values. The output values are managed via shared pointers.
+ * When a task subscribes to another task's results, it gets a copy of the shared pointer to the
+ * respective result. The producer task uses the shared pointer to populate the values and
+ * clears the outputs and inputs once done. This guarantees that the outputs don't live longer
+ * than necessary.
+ *
+ *
+ * Succeed vs Consume
+ * Succeed simply sets a dependency. It can take any Waitable (not just Task) and, importantly,
+ * it can take tasks that have been already submitted or even completed.
+ * Consume, by contrast, requires that the producer task is not yet submitted for execution. This
+ * limitation is necessary because of the output lifecycle.
+ */
+class Task : public CompletionEvent {
+  template <typename F, typename... Args>
+  void SetResult(F &&f, Args &&...args) {
+    assert(state == TaskState::Running);
+    assert(results_.size() == 1 && results_[0]);
+    results_[0]->SetResultOf([&]() { return f(std::forward<Args>(args)...); });
+  }
+
+  template <int i, typename... T>
+  static void UnpackResults(TaskResults &r, std::tuple<T...> &&t) {
+    if constexpr (i < sizeof...(T)) {
+      r[i]->Set(std::move(std::get<i>(t)));
+      UnpackResults<i + 1>(r, std::move(t));
+    }
+  }
+
+  template <typename F, typename... Args>
+  auto SetResults(F &&f, Args &&...args) {
+    assert(state == TaskState::Running);
+    try {
+      using result_t = std::remove_reference_t<decltype(f(std::forward<Args>(args)...))>;
+      if constexpr (detail::is_iterable_v<result_t>) {
+        auto &&results = f(std::forward<Args>(args)...);
+        size_t n = 0;
+        for (auto &&r : results) {
+          if (n >= results_.size())
+            throw std::logic_error("The function provided more results than "
+                                  "the task was declared to have.");
+          using T = std::remove_reference_t<decltype(r)>;
+          results_[n]->Set(std::forward<T>(r));
+          n++;
+        }
+
+        if (n < results_.size())
+          throw std::logic_error("The function provided fewer results than "
+                                  "the task was declared to have.");
```

Comment:
Can't we just comapre `results.size` with `results_.size()` at the begingin?

Reply:
It doesn't need to have .size(). I only require it to be iterable (a C array reference, a list, a range of iterators - all of that works).

Reply:
ok